### PR TITLE
[Bugfix] Fix showing buildlogs for Swift and Python on Bamboo and Jenkins

### PR DESCRIPTION
**Pull request by derLalla**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context

Build logs on Bamboo were not shown for Swift and Python programming exercises on Jenkins and Bamboo.

### Description

This PR enables to show the build logs for Swift and Python programming exercises on Jenkins and Bamboo.
Fetching build logs in Bamboo differed from fetching in Jenkins, as they were blacklisted in Bamboo and whitelisted in Jenkins. This is solved by providing filter functions ("isWarningLog", "isErrorLog", etc.) in the build log service which is blacklisting build logs for both CI systems.

Furhermore the Jenkins Pipeline and Bamboo Build Scripts for Python were adapted to be able to show build logs at all. The python code is now compiled at first and `pytest` gets executed only on success.

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a Swift or a Python Exercise
4. Open the Online Editor "Edit in Editor"
5. Introduce a compilation error (e.g. remove a '{')
6. Verify that the build logs are shown

Note: This PR needs to be testes on ts1/3 (Bamboo) and ts2 (Jenkins)

### Test Coverage

No new functionality was added so the test cases stay the same.

### Screenshots

Python:
Buildlog in Detail View:
image

Build Output:
image

Swift:
Buildlog in Detail View:
image

SCA Feedback shows in Detail View, despite build error (with SCA enabled):
image

Build Output:
image

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/BuildLogEntryService.java [hunk 1]
```diff
@@ -58,4 +59,72 @@ public BuildLogEntryService(BuildLogEntryRepository buildLogEntryRepository, Pro
...
     }
+
+    /**
+     * Determines if a given build log string can be categorized as an illegal reflection
+     *
+     * @param logString current build log string
+     * @return boolean indicating an illegal reflection log or not
+     */
+    public boolean isIllegalReflectionLog(String logString) {
+        return logString.startsWith("WARNING") && (logString.contains("An illegal reflective access operation has occurred")
```
Comment (ivanchimeno) [inline src/main/java/de/tum/in/www1/artemis/service/BuildLogEntryService.java:70]: Maybe replace the lines `logString.startsWith("WARNING")` and do an early exit if  the log isn't a warning:
```
if (!isWarningLog(log))
  return false;

// The rest here
return logString.contains ...
```

**Review by ivanchimeno — CHANGES_REQUESTED** [review]
Very nice refactoring :) Just left a small comment to further improve the code

**Review by simonzilker — CHANGES_REQUESTED** [review]
Tested with @marvinchimeka @JohannesWeiss @airelawaleria  on ts2 and ts3. For python exercises the build output was not shown on both servers.

**Review by marvinchimeka — CHANGES_REQUESTED** [review]
Confirming what @simonzilker said

**Review by JohannesWeiss — CHANGES_REQUESTED** [review]
Tested in testing session, can confirm the issue mentioned by @simonzilker  with the python exercises on both test servers.

**Review by airelawaleria — CHANGES_REQUESTED** [review]
What @simonzilker said

File: src/main/java/de/tum/in/www1/artemis/service/BuildLogEntryService.java (hunk 1 above)
Comment (derLalla) [reply src/main/java/de/tum/in/www1/artemis/service/BuildLogEntryService.java:70]: this is actually a good idea, if all logs needed to start with "[WARNING]" to fall into this category. 
But in this specific case the log either starts with "WARNING..." or it just contains `"to enable warnings of further illegal reflective access operations"` or `"All illegal access operations will be denied in a future release"`

So this change would show the logs.

**Comment by derLalla** [comment]
> Tested with @marvinchimeka @JohannesWeiss @airelawaleria on ts2 and ts3. For python exercises the build output was not shown on both servers.

can you tell me on which server you created which exercise, so I can have a look into the generated buildplan?

**Comment by derLalla** [comment]
> > Tested with @marvinchimeka @JohannesWeiss @airelawaleria on ts2 and ts3. For python exercises the build output was not shown on both servers.
> 
> can you tell me on which server you created which exercise, so I can have a look into the generated buildplan?

I've tested it again on ts2 -> https://artemistest2.ase.in.tum.de/course-management/5/programming-exercises/2647/code-editor/9933
Changed the Template to not compile and it worked as expected.

File: src/main/java/de/tum/in/www1/artemis/service/BuildLogEntryService.java (hunk 1 above)
Comment (ivanchimeno) [reply src/main/java/de/tum/in/www1/artemis/service/BuildLogEntryService.java:70]: Good point

**Review by airelawaleria — APPROVED** [review]
Tested on ts1, ts2. Approve

**Review by simonzilker — APPROVED** [review]
Retested and did not face the issue anymore. Sorry! Approve!

**Review by LudwigWeide — APPROVED** [review]
Approve.

**Review by marvinchimeka — APPROVED** [review]
Approve. Works as intended.

**Review by DenizShabani — APPROVED** [review]
Works as intended!

**Review by JohannesWeiss — APPROVED** [review]
Approve as well!

**Review by EneaGore — APPROVED** [review]
Buildlogs look good on both severs. On ts1 it seemed the buildlogs were longer and had more information.

**Review by kloessst — CHANGES_REQUESTED** [review]
Tested with multiple exercises on TS1 and TS2. Worked fine. Just one question though

File: src/main/java/de/tum/in/www1/artemis/service/BuildLogEntryService.java [hunk 2]
```diff
@@ -58,4 +59,72 @@ public BuildLogEntryService(BuildLogEntryRepository buildLogEntryRepository, Pro
...
+     * @return boolean indicating a build log should be added to the overall build logs
+     */
+    public boolean checkIfBuildLogIsNotADuplicate(List<BuildLogEntry> buildLogs, String shortenedLogString) {
+        // E.g. Swift produces a lot of duplicate build logs when a build fails
+        if (!buildLogs.isEmpty()) {
+            var existingLog = buildLogs.stream().filter(log -> log.getLog().equals(shortenedLogString)).findFirst();
+            String lastLog = buildLogs.get(buildLogs.size() - 1).getLog();
+            // If the log does not exist already or if the log is a single blank log add it to the build logs (avoid more than one empty log in a row)
+            boolean isSingleBlankLog = shortenedLogString.isBlank() && !lastLog.isBlank();
+            return existingLog.isEmpty() || isSingleBlankLog;
```
[the remark this answers is not in the document]
Comment (derLalla) [reply src/main/java/de/tum/in/www1/artemis/service/BuildLogEntryService.java:126]: these are good for the visual improvement of the logs.  
And mostly there are added by the respective compile message itself so I didn't want to remove them.

I noticed that for SwiftLint some messages are duplicated very often (for each File the same message) but blank messages should be excluded from that logic.

**Review by ivanchimeno — APPROVED** [review]

**Review by krusche — APPROVED** [review]

**Pull request merged**